pallet/staking/src/lib.rs

```diff
@@ -735,215 +827,276 @@ pub mod pallet {
+								.fold(0, |r, d| r + T::Deposit::amount(who, d).unwrap_or_default()),
+					) + balance2power::<KtonPool<T>>(l.staked_kton)
+				})
+				.unwrap_or_default()
 		}
 
 		/// Calculate the power of the given account.
-		pub fn power_of(who: &T::AccountId) -> Power {
+		///
+		/// This is an optimized version of [`Self::quick_power_of`].
+		/// Avoiding read the pools' storage multiple times.
+		pub fn power_of(who: &T::AccountId, ring_pool: Balance, kton_pool: Balance) -> Power {
+			const HALF_POWER: u128 = 500_000_000;
+
 			<Ledgers<T>>::get(who)
 				.map(|l| {
-					Self::balance2power::<RingPool<T>>(
+					(Perquintill::from_rational(
 						l.staked_ring
 							+ l.staked_deposits
 								.into_iter()
 								// We don't care if the deposit exists here.
 								// It was guaranteed by the `stake`/`unstake`/`restake` functions.
 								.fold(0, |r, d| r + T::Deposit::amount(who, d).unwrap_or_default()),
-					) + Self::balance2power::<KtonPool<T>>(l.staked_kton)
+						ring_pool.max(1),
+					) * HALF_POWER)
+						.saturating_add(
+							Perquintill::from_rational(l.staked_kton, kton_pool.max(1))
+								* HALF_POWER,
+						) as _
 				})
 				.unwrap_or_default()
 		}
 
-		// TODO: weight
-		/// Pay the session reward to the stakers.
-		pub fn payout(amount: Balance) -> Balance {
-			let (total_points, reward_map) = <RewardPoints<T>>::get();
-			// Due to the `payout * percent` there might be some losses.
-			let mut actual_payout = 0;
-
-			for (c, p) in reward_map {
-				let Some(commission) = <Collators<T>>::get(&c) else {
-						#[cfg(test)]
-						panic!("[pallet::staking] collator({c:?}) must be found; qed");
-						#[cfg(not(test))]
-						{
-							log::error!("[pallet::staking] collator({c:?}) must be found; qed");
-
-							continue;
-						}
-					};
-				let c_total_payout = Perbill::from_rational(p, total_points) * amount;
-				let mut c_payout = commission * c_total_payout;
-				let n_payout = c_total_payout - c_payout;
-				let Some(c_exposure) = <Exposures<T>>::get(&c) else {
-						#[cfg(test)]
-						panic!("[pallet::staking] exposure({c:?}) must be found; qed");
-						#[cfg(not(test))]
-						{
-							log::error!("[pallet::staking] exposure({c:?}) must be found; qed");
-
-							continue;
-						}
-					};
+		/// Pay the session reward to staking pot and update the stakers' reward record.
+		pub fn dispatch_session_reward(amount: Balance) -> Balance {
+			let (sum, map) = <AuthoredBlocksCount<T>>::take();
+			let staking_pot = account_id();
+			let mut actual_reward = 0;
+			let mut unpaid = 0;
 
-				for n_exposure in c_exposure.nominators {
-					let n_payout =
-						Perbill::from_rational(n_exposure.vote, c_exposure.vote) * n_payout;
+			map.into_iter().for_each(|(c, p)| {
+				let r = Perbill::from_rational(p, sum) * amount;
 
-					if c == n_exposure.who {
-						// If the collator nominated themselves.
+				<Unpaid<T>>::mutate(&c, |u| *u = u.map(|u| u + r).or(Some(r)));
```

Comment:
Following the latest design, I think this line should be moved down to the `T::InflationManager::reward(&staking_pot, r).is_ok()` branch. If the inflation manager reward failture fails, the collator does not get any token reward. No unpaid.

Reply:
Renaming is acceptable, but it should still be counted even if the reward fails.

Reply:
>  it should still be counted even if the reward fails.

In the case of reward fails, the actual reward that the collator receives is less than the PendingReward value, in the next session, the collator may lose their original fund to pay out the nominator's reward. Is this expected behaviour?

Reply:
Yep